.github/workflows/build.yml

```diff
@@ -0,0 +1,102 @@
+env:
+  # Use docker.io for Docker Hub if empty
+  REGISTRY: ghcr.io
+  # github.repository as <account>/<repo>
+  IMAGE_NAME: ${{ github.repository }}
+
+jobs:
+  build_matrix:
+    strategy:
+      matrix:
+        # Maybe we want to have executables for all platforms just to test this?
+        os: [ubuntu-latest]
+        rust: [stable]
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up cache
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+            target/
+          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+      - uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+      - uses: actions-rs/cargo@v1
+        with:
+          command: build
+          args: --release --all-features
+      - shell: bash
+        run: |
+          mkdir artifacts
+          mv ./target/release/cotevalentines ./artifacts
+      - name: Run UPX
+        uses: crazy-max/ghaction-upx@v2
+        with:
+          version: latest
+          files: |
+            ./artifacts/cotevalentines
+          # args: -fq
+      - uses: actions/upload-artifact@v3
+        name: Upload Binary
+        with:
+          name: Cotevalentines2023
+          path: ./artifacts/cotevalentines*
+  build_docker_image:
+    runs-on: ubuntu-latest
+    needs: build_matrix
+    permissions:
+      contents: read
+      packages: write
+    if: github.ref == 'refs/heads/release' || github.ref == 'refs/heads/docker'
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: actions/download-artifact@v2
+        name: Download release artifact
+        with:
+          name: Cotevalentines2023
+          path: ./docker/Cotevalentines2023
+
+      - shell: bash
+        name: Copy Dockerfile
+        run: |
+          cp ./Dockerfile ./docker
+
+      - name: Setup Docker buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Log into registry ${{ env.REGISTRY }}
+        if: github.event_name != 'pull_request'
+        uses: docker/login-action@v2
+        with:
+          registry: ${{ env.REGISTRY }}
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      
+      # Extract metadata (tags, labels) for Docker
+      # https://github.com/docker/metadata-action
+      - name: Extract Docker metadata
+        id: meta
+        uses: docker/metadata-action@v4.3.0
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
+
+      # Build and push Docker image with Buildx (don't push on PR)
+      # https://github.com/docker/build-push-action
+      - name: Build and push Docker image
+        id: build-and-push
+        uses: docker/build-push-action@v4
+        with:
+          context: ./docker
+          push: ${{ github.event_name != 'pull_request' }}
```

Comment:
I was about to say this shouldn't push but I'm literally working on the branch itself of course.